### PR TITLE
feat: action handler for joining conversation with key and code #SQCORE-1134

### PR DESCRIPTION
**Pull request by sb88k**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQCORE-1134" title="SQCORE-1134" target="_blank">TaskSQCORE-1134</a>  [iOS] Refactor join conversation by code to use action pattern
  </td></table>
  <br />
 

## What's new in this PR?

Adding an action handler for joining conversation

## Dependencies

- https://github.com/wireapp/wire-ios-data-model/pull/1267

### CODE REVIEW

File: Sources/Request Strategies/Conversation/Actions/JoinActionHandlerTests.swift [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import XCTest
+@testable import WireRequestStrategy
+
+final class JoinActionHandlerTests: MessagingTestBase {
```
Comment (typfel) [inline Sources/Request Strategies/Conversation/Actions/JoinActionHandlerTests.swift:21]: Let's add a test for the error code / label mapping. 

An example of a similar test can be found here https://github.com/wireapp/wire-ios-request-strategy/blob/f1d0ab559a0d146f4a5dd35d6f6515e8d4ec5945/Sources/Request%20Strategies/Connection/Actions/ConnectToUserActionHandlerTests.swift#L76

**Review by typfel — CHANGES_REQUESTED** [review]

File: Cartfile.resolved [hunk 2]
```diff
@@ -4,7 +4,11 @@ github "wireapp/Starscream" "4.0.4_xcode13"
 github "wireapp/ocmock" "v3.4.3_xcframework"
 github "wireapp/swift-protobuf" "1.15.0_xcframework"
 github "wireapp/wire-ios-cryptobox" "31.0.0"
-github "wireapp/wire-ios-data-model" "276.0.0"
+<<<<<<< HEAD
+github "wireapp/wire-ios-data-model" "9f9cec695d8220f99ad9ad8306752a11903b02de"
+=======
+github "wireapp/wire-ios-data-model" "45207dd2a0c8edf6e587bb18433da90d9325bb8d"
+>>>>>>> 7517667 (apply suggestions)
```
Comment (KaterinaWire) [inline Cartfile.resolved:11]:
```suggestion
github "wireapp/wire-ios-data-model" "feature/join-conversation-action-handler-SQCORE-1134"

```

Comment (KaterinaWire) [inline Cartfile.resolved:11]: I think it was left after the merging. Should be:
github "wireapp/wire-ios-data-model" "feature/join-conversation-action-handler-SQCORE-1134"

File: Sources/Request Strategies/Conversation/Actions/JoinActionHandler.swift [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        var action = action
+        
+        switch response.httpStatus {
+        case 200:
+            guard
+                let payload = response.payload,
+                let event = ZMUpdateEvent(fromEventStreamPayload: payload, uuid: nil),
+                let rawData = response.rawData,
+                let conversationEvent = Payload.ConversationEvent<Payload.UpdateConverationMemberJoin>(rawData, decoder: .defaultDecoder),
+                let conversationID = event.payload["conversation"] as? String
```
[the remark this answers is not in the document]
Comment (sb88k) [reply Sources/Request Strategies/Conversation/Actions/JoinActionHandler.swift]: Done

File: Sources/Request Strategies/Conversation/Actions/JoinConversationActionHandlerTests.swift [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            let expectation = self.expectation(description: "Result Handler was called")
+            
+            var action = JoinConversationAction(key: key, code: code)
+            action.onResult { result in
+                if case .failure = result {
+                    expectation.fulfill()
+                }
+            }
+            
+            let response = ZMTransportResponse(payload: nil, httpStatus: 200, transportSessionError: nil)
```
Comment (KaterinaWire) [inline Sources/Request Strategies/Conversation/Actions/JoinConversationActionHandlerTests.swift:115]: But the error code in the name of the test is 204. This test should fail.

File: Sources/Request Strategies/Conversation/Actions/JoinConversationActionHandler.swift [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        guard
+            let payloadData = payload.payloadData(encoder: .defaultEncoder),
+            let payloadAsString = String(bytes: payloadData, encoding: .utf8)
+        else {
+            var action = action
+            action.notifyResult(.failure(.unknown))
+            return nil
+        }
+        
+
```
Comment (KaterinaWire) [inline Sources/Request Strategies/Conversation/Actions/JoinConversationActionHandler.swift:48]: An extra empty line

**Pull request closed without merging**